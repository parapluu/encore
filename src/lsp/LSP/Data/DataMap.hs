module LSP.Data.DataMap(
  LSPData,
  DataMap,
  hasErrorInDataMap
  )
where

-- ###################################################################### --
-- Section: Imports
-- ###################################################################### --

-- Standard
import Data.Map as Map hiding (foldr)

-- LSP
import LSP.Data.Program
import LSP.Data.TextDocument

-- ###################################################################### --
-- Section: Data
-- ###################################################################### --

type LSPData = (Program, TextDocument)

type DataMap = Map String LSPData

-- ###################################################################### --
-- Section: Functions
-- ###################################################################### --

hasErrorInDataMap :: DataMap -> Bool
hasErrorInDataMap dataMap = Map.foldl (\acc (prog, _) -> acc || (length (errors prog)) > 0) False dataMap
