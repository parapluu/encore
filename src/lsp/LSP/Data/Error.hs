module LSP.Data.Error (
  Error(..),
  ErrorType(..),
  fromTCError,
  fromTCErrors,
  fromTCWarning,
  fromTCWarnings,
  fromParsecError,
  fromErrorMessage
) where

-- ###################################################################### --
-- Section: Imports
-- ###################################################################### --

-- Library
import Text.Megaparsec

-- Standard
import qualified Data.Map.Strict as Map
import qualified Data.List.NonEmpty as NE(head)
import Data.List
import Control.Monad

-- Encore imports
import Typechecker.TypeError hiding (Error)
import AST.AST
import qualified AST.Meta as ASTMeta(Position(SingletonPos, RangePos))

-- ###################################################################### --
-- Section: Data
-- ###################################################################### --

{- Position tuple  -}
type Position = ((Int, Int), (Int, Int))

data ErrorType = TError | TWarning | THint

instance Show ErrorType where
  show (TError) = "Error"
  show (TWarning) = "Warning"
  show (THint) = "Hint"

{- Error datastructure. Can represent a variety of different error types -}
data Error = Error{
    message     :: String,      -- Error message string
    position    :: Position,    -- Start and end position
    errorType   :: ErrorType
} deriving (Show)

-- ###################################################################### --
-- Section: Functions
-- ###################################################################### --

extractTCErrorPosition :: TCError -> Position
extractTCErrorPosition error@(TCError errorType []) = ((0, 0), (0, 0))
extractTCErrorPosition error@(TCError errorType backtrace) =
    case fst (head backtrace) of
        (ASTMeta.SingletonPos startPos) ->
            ((getPosLine startPos - 1, getPosCol startPos - 1),
            (getPosLine startPos - 1, getPosCol startPos - 1))
        (ASTMeta.RangePos startPos endPos) ->
            ((getPosLine startPos - 1, getPosCol startPos - 1),
            (getPosLine endPos - 1, getPosCol endPos - 1))
    where getPosLine = fromIntegral . unPos . sourceLine
          getPosCol  = fromIntegral . unPos . sourceColumn

extractTCWarningPosition :: TCWarning -> Position
extractTCWarningPosition warning = ((0,0), (0,0))

fromTCError :: TCError -> Error
fromTCError error =
    Error{message = show error, position = extractTCErrorPosition error, errorType = TError}

fromTCErrors :: [TCError] -> [Error]
fromTCErrors errors = fmap (fromTCError) errors

fromTCWarning :: TCWarning -> Error
fromTCWarning warning =
    Error{message = show warning, position = extractTCWarningPosition warning, errorType = TWarning}

fromTCWarnings :: [TCWarning] -> [Error]
fromTCWarnings warnings = fmap (fromTCWarning) warnings

fromParsecError :: (ParseError Char Text.Megaparsec.Dec) -> Error
fromParsecError error =
    Error{
        message = show error,
        position = ((fromIntegral(unPos $ sourceLine $ NE.head $ errorPos error), fromIntegral(unPos $ sourceColumn $ NE.head $ errorPos error)),
                    (fromIntegral(unPos $ sourceLine $ NE.head $ errorPos error), fromIntegral(unPos $ sourceColumn $ NE.head $ errorPos error))),
        errorType = TError
    }

{- Construct a Error from a message, position and whether it is actually a
    warning

    Param: Error message
    Param: Position (row, column)
    Param: Whether or not the error is actually a warning
-}
fromErrorMessage :: String -> Position -> ErrorType -> Error
fromErrorMessage _message _position _errorType =
    Error{message = _message, position = _position, errorType = _errorType}
