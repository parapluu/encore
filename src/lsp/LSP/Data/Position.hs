module LSP.Data.Position (
  Position,
  Range,
  inRange,
  fromSourcePos,
  fromSourcePosRange,
  rangeCompliment,
  widestRange
) where

-- ###################################################################### --
-- Section: Imports
-- ###################################################################### --

-- Library
import Text.Megaparsec(SourcePos,unPos,sourceLine,sourceColumn)
import Data.List as List

-- ###################################################################### --
-- Section: Data
-- ###################################################################### --

{-  -}
type Position = (Int, Int)

{-  -}
type Range = (Position, Position)

-- ###################################################################### --
-- Section: Functions
-- ###################################################################### --

{-  -}
inRange :: Position -> Range -> Bool
inRange pos (start, end)
  | fst pos > fst start && fst pos < fst end = True
  | fst pos == fst start && fst pos < fst end && snd pos >= snd start = True
  | fst pos > fst start && fst pos == fst end && snd pos < snd end = True
  | fst pos == fst start && fst pos == fst end && snd pos >= snd start && snd pos < snd end = True
  | True = False -- sorry, had to do it (otherwise == True)

{-  -}
fromSourcePos :: SourcePos -> Position
fromSourcePos pos = (fromIntegral (unPos (sourceLine pos)), fromIntegral (unPos (sourceColumn pos)))

{-  -}
fromSourcePosRange :: SourcePos -> SourcePos -> Range
fromSourcePosRange from to = (fromSourcePos from, fromSourcePos to)

posLess :: Position -> Position -> Bool
posLess a b
  | fst a < fst b = True
  | fst a == fst b && snd a < snd b = True
  | otherwise = False

posGreater :: Position -> Position -> Bool
posGreater a b
  | fst a > fst b = True
  | fst a == fst b && snd a > snd b = True
  | otherwise = False

rangeCompliment :: Range -> Range -> Range
rangeCompliment a b
  | posLess (fst a) (fst b) = (fst a, fst b)
  | posGreater (snd a) (snd b) = (snd b, snd a)
  | otherwise = a

minPosition :: [Position] -> Position
minPosition [] = (0, 0)
minPosition (x:[]) = x
minPosition (x:y:xs)
    | posLess x y = minPosition (x:xs)
    | otherwise = minPosition (y:xs)

maxPosition :: [Position] -> Position
maxPosition [] = (0, 0)
maxPosition (x:[]) = x
maxPosition (x:y:xs)
    | posGreater x y = maxPosition (x:xs)
    | otherwise = maxPosition (y:xs)

widestRange :: Range -> [Range] -> Range
widestRange firstRange ranges = do
  let wholeRange = (firstRange:ranges)
  (minPosition (fmap fst wholeRange), maxPosition (fmap snd wholeRange))
