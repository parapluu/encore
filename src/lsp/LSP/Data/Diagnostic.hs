{-# LANGUAGE OverloadedStrings #-}

module LSP.Data.Diagnostic (
    PublishDiagnosticsParams(..),
    Diagnostic(..),
    DiagnosticSeverity(..),
    errorToDiagnostic
) where

import Data.Aeson

import LSP.Data.Position
import LSP.Data.Error

data PublishDiagnosticsParams = PublishDiagnosticsParams {
    pdpUri :: String,
    pdpDiagnostics :: [Diagnostic]
} deriving (Show)

data Diagnostic = Diagnostic {
    dRange :: Range,
    dSeverity :: DiagnosticSeverity,
    dCode :: String,
    dSource :: String,
    dMessage :: String
} deriving (Show)

data DiagnosticSeverity
    = DError
    | DWarning
    | DInformation
    | DHint
    deriving (Show, Enum)

errorToDiagnostic :: Error -> Diagnostic
errorToDiagnostic error =
    Diagnostic {
        dRange    = position error,
        dCode     = "-- error --",
        dSource   = "Encore",
        dMessage  = message error,
        dSeverity = case errorType error of
            TError   -> DError
            TWarning -> DWarning
            THint    -> DHint
    }

{-instance Enum DiagnosticSeverity where
    toEnum = fromIntegral . (-1)
    fromEnum = (+1) . fromInteger-}

instance ToJSON PublishDiagnosticsParams where
    toJSON pdp =
        object [
                "uri"         .= pdpUri pdp,
                "diagnostics" .= pdpDiagnostics pdp
            ]

instance ToJSON Diagnostic where
    toJSON diagnostic =
        object [
                "range"    .= object [
                        "start" .= object [
                                "line"      .= (fst $ fst $ dRange diagnostic),
                                "character" .= (snd $ fst $ dRange diagnostic)
                            ],
                        "end"   .= object [
                                "line"      .= (fst $ snd $ dRange diagnostic),
                                "character" .= (snd $ snd $ dRange diagnostic)
                            ]
                    ],
                "severity" .= (1 + (fromEnum $ dSeverity diagnostic)),
                "code"     .= dCode diagnostic,
                "source"   .= dSource diagnostic,
                "message"  .= dMessage diagnostic
            ]