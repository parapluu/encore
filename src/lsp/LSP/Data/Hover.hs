{-# LANGUAGE OverloadedStrings #-}

module LSP.Data.Hover where

-- ###################################################################### --
-- Section: Imports
-- ###################################################################### --
import LSP.Data.Position
import Data.Aeson

-- ###################################################################### --
-- Section: Data
-- ###################################################################### --
data TextDocumentPositionParams = TextDocumentPositionParams {
    uri :: String,
    position :: Position
}

data Hover = Hover {
    contents :: String,
    range :: Range
}

-- ###################################################################### --
-- Section: Type Classes
-- ###################################################################### --

instance FromJSON TextDocumentPositionParams where
    parseJSON = withObject "params" $ \o -> do
        textDocument    <- o .: "textDocument"
        uri             <- textDocument .: "uri"
        position        <- o .: "position"
        line            <- position .: "line"
        character       <- position .: "character"

        return TextDocumentPositionParams {
            uri = uri,
            position = (line + 1, character + 1)
        }

instance ToJSON Hover where
    toJSON hover =
        object [
            "contents" .= contents hover,
            "range"    .= object [
                "start" .= object [
                    "line"      .= ((fst $ fst $ range hover) - 1),
                    "character" .= ((snd $ fst $ range hover) - 1)
                ],
                "end"   .= object [
                    "line"      .= ((fst $ snd $ range hover) - 1),
                    "character" .= ((snd $ snd $ range hover) - 1)
                ]
            ]
        ]
