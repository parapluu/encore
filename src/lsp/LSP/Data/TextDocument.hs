{-# LANGUAGE OverloadedStrings #-}

module LSP.Data.TextDocument (
    TextDocument(..),
    TextDocumentClose(..),
    TextDocumentChange(..),
    TextDocumentIdent(..),
    uri,
    version,
    applyTextDocumentChange,
    makeBlankTextDocument
) where

-- ###################################################################### --
-- Section: Imports
-- ###################################################################### --

-- Standard
import Data.Aeson
import Data.String.Utils (split, join)

-- LSP
import LSP.Data.Position

-- ###################################################################### --
-- Section: Data
-- ###################################################################### --

data TextDocument = TextDocument {
    tdUri :: String,
    tdVersion :: Int,
    tdLanguageId :: String,
    tdContents :: String
} deriving (Show)

data TextDocumentClose = TextDocumentClose {
    tdclIdentifier :: TextDocumentIdent
} deriving (Show)

data TextDocumentChange = TextDocumentChange {
    tdcIdentifier :: TextDocumentIdentVersion,
    tdcChanges :: [TextDocumentContentChange]
} deriving (Show)

data TextDocumentIdent = TextDocumentIdent {
    tdiUri :: String
} deriving (Show)

data TextDocumentIdentVersion = TextDocumentIdentVersion {
    tdivUri :: String,
    tdivVersion :: Int
} deriving (Show)

data TextDocumentContentChange = TextDocumentContentChange {
    tdccRange :: Range,
    tdccRangeLength :: Int,
    tdccText :: String
} deriving (Show)

-- ###################################################################### --
-- Section: Functions
-- ###################################################################### --

applyTextDocumentChange :: TextDocumentChange -> TextDocument -> TextDocument
applyTextDocumentChange textDocumentChange textDocument =
    foldr (\change textDocument ->
            TextDocument {
                tdUri        = tdUri textDocument,
                tdVersion    = tdivVersion $ tdcIdentifier textDocumentChange,
                tdLanguageId = tdLanguageId textDocument,
                tdContents   = applyTextChange (tdccRange change)
                                               (tdccText change)
                                               (tdContents textDocument)
            }
        ) textDocument $ tdcChanges textDocumentChange

applyTextChange :: Range -> String -> String -> String
applyTextChange _ replacement "" = replacement
applyTextChange a@((startLine, startChar), (endLine, endChar))
                replacement
                text
    = let textLines = split "\n" text
          clamp = max 0 . min (length textLines - 1)
          boundedStartLine = clamp startLine
          boundedEndLine   = clamp endLine
          startSegment = join "\n" $ take startLine textLines ++
                                     [take startChar $ textLines !! boundedStartLine]
          endSegment = join "\n" $ (drop endChar $ textLines !! boundedEndLine) :
                                   drop (endLine+1) textLines
      in startSegment ++ replacement ++ endSegment

makeBlankTextDocument :: String -> TextDocument
makeBlankTextDocument name = TextDocument{
    tdUri      = name,
    tdVersion  = 1,
    tdLanguageId = "encore",
    tdContents   = ""
}

-- ###################################################################### --
-- Section: Type Classes
-- ###################################################################### --

class TextDocumentURI a where
    uri :: a -> String
class TextDocumentVersion a where
    version :: a -> Int

instance TextDocumentURI TextDocument where
    uri = tdUri
instance TextDocumentVersion TextDocument where
    version = tdVersion

instance TextDocumentURI TextDocumentChange where
    uri = uri . tdcIdentifier
instance TextDocumentVersion TextDocumentChange where
    version = version . tdcIdentifier

instance TextDocumentURI TextDocumentIdent where
    uri = tdiUri

instance TextDocumentURI TextDocumentIdentVersion where
    uri = tdivUri
instance TextDocumentVersion TextDocumentIdentVersion where
    version = tdivVersion

instance FromJSON TextDocument where
    parseJSON = withObject "params" $ \o -> do
        document <- o .: "textDocument"
        uri      <- document .: "uri"
        version  <- document .: "version"
        id       <- document .: "languageId"
        text     <- document .: "text"
        return TextDocument {
            tdUri = uri,
            tdVersion = version,
            tdLanguageId = id,
            tdContents = text
        }

instance FromJSON TextDocumentClose where
    parseJSON = withObject "params" $ \o -> do
        identifier <- o .: "textDocument"

        return TextDocumentClose {
            tdclIdentifier = identifier
        }

instance FromJSON TextDocumentChange where
    parseJSON = withObject "params" $ \o -> do
        identifier  <- o .: "textDocument"
        changes     <- o .: "contentChanges"

        return TextDocumentChange {
            tdcIdentifier = identifier,
            tdcChanges = changes
        }

instance FromJSON TextDocumentContentChange where
    parseJSON = withObject "TextDocumentChange" $ \o -> do
        range       <- o .: "range"
        rangeStart  <- range .: "start"
        startLine   <- rangeStart .: "line"
        startChar   <- rangeStart .: "character"
        rangeEnd    <- range .: "end"
        endLine     <- rangeEnd .: "line"
        endChar     <- rangeEnd .: "character"
        rangeLength <- o .: "rangeLength"
        text        <- o .: "text"

        return TextDocumentContentChange {
            tdccRange = ((startLine, startChar), (endLine, endChar)),
            tdccRangeLength = rangeLength,
            tdccText = text
        }

instance FromJSON TextDocumentIdent where
    parseJSON = withObject "textDocumentIdentifier" $ \o -> do
        uri <- o .: "uri"
        return TextDocumentIdent {
            tdiUri = uri
        }

instance FromJSON TextDocumentIdentVersion where
    parseJSON = withObject "textDocumentIdentifier" $ \o -> do
        uri <- o .: "uri"
        version <- o .: "version"
        return TextDocumentIdentVersion {
            tdivUri = uri,
            tdivVersion = version
        }
