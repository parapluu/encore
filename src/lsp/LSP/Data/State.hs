module LSP.Data.State (
    LSPState(..),
    initial,
    lookupTextDocument,
    addTextDocument,
    closeTextDocument,
    changeTextDocument,
    produceTextDocument,
    compileDocument,
    getProgram
) where

-- ###################################################################### --
-- Section: Imports
-- ###################################################################### --

-- Standard
import Data.Map as Map hiding (foldr)

-- LSP
import LSP.Data.TextDocument
import LSP.Data.Program
import LSP.Data.DataMap
import LSP.Producer

-- ###################################################################### --
-- Section: Data
-- ###################################################################### --

data LSPState = LSPState {
    programs :: DataMap
} deriving (Show)

-- ###################################################################### --
-- Section: Functions
-- ###################################################################### --

initial :: LSPState
initial = LSPState {
    programs = Map.empty
}

lookupTextDocument :: String -> LSPState -> Maybe (TextDocument)
lookupTextDocument uri state =
  case Map.lookup uri (programs state) of
    Just res -> Just $ snd res
    Nothing -> Nothing

addTextDocument :: TextDocument -> LSPState -> LSPState
addTextDocument newDocument (LSPState programs) =
    LSPState (Map.insert (uri newDocument) (makeBlankProgram (uri newDocument), newDocument) programs)

closeTextDocument :: TextDocumentIdent -> LSPState -> LSPState
closeTextDocument ident = LSPState . Map.delete (uri ident) . programs

changeTextDocument :: TextDocumentChange -> LSPState -> LSPState
changeTextDocument documentChange state@(LSPState programs) =
    case Map.lookup (uri documentChange) programs of
        Nothing -> state
        Just (program, textDocument) ->
            LSPState $ Map.insert (uri documentChange)
                                  (program, applyTextDocumentChange documentChange textDocument)
                                  programs

produceTextDocument :: TextDocument -> LSPState -> IO (LSPState)
produceTextDocument textDocument state = do
  let path = uri textDocument
  dataMap <- produceAndUpdateState path (programs state)
  return (LSPState {programs = dataMap})

compileDocument :: String -> LSPState -> IO LSPState
compileDocument path = fmap LSPState . produceAndUpdateState path . programs

getProgram :: String -> LSPState -> Maybe Program
getProgram path = fmap fst . Map.lookup path . programs
