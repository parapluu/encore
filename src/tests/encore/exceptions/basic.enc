import Exceptional
import Data.Either

fun test_tryCatch(c : int) : unit
  try
    println(c)
  catch
    println(-c)
  end
end

fun test_tryCatch_explicit(c : int) : unit
  try
    println(c)
  catch RuntimeException
    println(-c)
  end
end

fun test_tryCatch_throw(c : int) : unit
  try
    throw RuntimeException("")
    println(-c)
  catch
    println(c)
  end
end

fun test_tryCatch_throw_explicit(c : int) : unit
  try
    throw RuntimeException("")
    println(-c)
  catch RuntimeException
    println(c)
  end
end


fun test_throw_without_msg(c : int) : unit
  try
    throw RuntimeException
    println(-c)
  catch
    println(c)
  end
end

fun test_tryFin(c : int) : unit
  var x = 0
  try
    x = x+c
  finally
    x *= 100
  end
  println(x)
end

fun test_tryCatchFin(c : int) : unit
  var x = 0
  try
    x = x+c
  catch
    x = x-c
  finally
    x = x*100
  end
  println(x)
end

fun test_tryCatchFin_throw(c : int) : unit
  var x = 0
  try
    throw RuntimeException
    x = x-c
  catch
    x = x+c
  finally
    x = x*100
  end
  println(x)
end

fun test_tryCatchFin_nested(c : int) : unit
  -- Should result in c*100
  var x = 0
  try
    try
      x = x+c
    catch
      x = x-c
    end
  catch
    x = 0
  finally
    x = x*100
  end
  println(x)
end

fun test_tryCatchFin_nested_throw(c : int) : unit
  -- Should result in c*100
  var x = 0
  try
    try
      throw RuntimeException
      x = x-c
    catch
      x = x+c
    end
  catch
    x = 0
  finally
    x = x*100
  end
  println(x)
end

fun deep_nesting_aux(current : int, accumulator : int) : int
  try
    if current > 0 then
      deep_nesting_aux(current-1, accumulator+current)
    else
      accumulator
    end
  catch e : RuntimeException
    -10
  finally
    ()
  end
end

fun test_deep_nesting() : unit
  println(deep_nesting_aux(100, 0))
end

fun test_tryFin_throw(c : int) : unit
  -- The nested tryFin won't catch the exception,
  -- but it should be caught in the outer scope.
  -- Should result in c*100
  var x = 0
  try
    try
      throw RuntimeException
      x = x-c
    finally
      x = x+100
    end
  catch
    x = c*x
  end
  println(x)
end

fun test_exception_var(c : int) : unit
  try
    throw RuntimeException
    println(-c)
  catch e : RuntimeException
    println(c)
  end
end

fun test_exception_msg() : unit
  var message = "informative_msg"
  try
    throw RuntimeException("informative_msg")
  catch e : RuntimeException
    println(e.message)
  end
end

fun test_exception_info() : unit
  try
    throw RuntimeException("informative_msg")
  catch e : RuntimeException
    println("type:{}, message:{}, line:{}, file:{}", e.type, e.message, "NOT_TESTING", "NOT_TESTING")
  end
end

exception FunException(RuntimeException)
fun test_catch_custom_throw(c : int) : unit
  try
    throw FunException("haha")
    println(-c)
  catch FunException
    println(c)
  end
end

fun test_catch_supertype(c : int) : unit
  -- If an exception type isn't specified in a catch branch,
  -- instances of that type should be caught in branches for supertypes.
  try
    throw FunException("haha")
    println(-c)
  catch RuntimeException
    println(c)
  end
end

exception CoolException(RuntimeException)
fun test_multiple_catch_branches() : unit
  try
    throw CoolException
  catch CoolException
    println("passed_multiple_catch_branches")
  catch
    println("failed_multiple_catch_branches")
  end
end

fun test_multiple_catch_branches2() : unit
  try
    throw RuntimeException
  catch CoolException
    println("failed_multiple_catch_branches2")
  catch
    println("passed_multiple_catch_branches2")
  end
end

exception LocalException(RuntimeException)
exception ReadException(RuntimeException)
exception LinearException(RuntimeException)
local class LocalClass
  def throw_exc() : unit
    throw LocalException
  end
end
read class ReadClass
  def throw_exc() : unit
    throw ReadException
  end
end
linear class LinearClass
  def throw_exc() : unit
    throw LinearException
  end
end
fun test_passive_classes_throw() : unit
  try
    try
      (new LocalClass).throw_exc()
    catch LocalException
      print("caught LocalException, ")
    end
    try
      (new ReadClass).throw_exc()
    catch ReadException
      print("ReadException, ")
    end
    try
      (new LinearClass).throw_exc()
    catch LinearException
      println("LinearException")
    end
  catch
    println("test failed")
  end    
end

fun test_rethrow() : unit
  var res = "fail"
  try
    throw RuntimeException("wahaa!")
    res = "fail1"
  catch e : RuntimeException
    try
      e.rethrow()
      res = "fail2"
    catch e2 : RuntimeException
      res = "pass"
    end
  end
  println("test_rethrow: {}", res)
end

fun test_nested_rethrow() : unit
  var res = 1
  var e = new Exception(EMBED(EMBED const e4c_exception_type_t* END) &RuntimeException; END, "fail", 0, "")
  var check = true
  try
    res = res+2
    try
      res = res+4
      throw RuntimeException("4")
    catch e_ : RuntimeException
      res = res+8
      e = e_
      e_.rethrow()
      ()
    end
  catch e_ : RuntimeException
    check = check && e_ == e
    res = res+16
  end
  println("{} {} {}", res, check, e.message)
end

linear class Acc
  var res : int
  var e : Exception
  var check : bool
  def init() : unit
    this.res = 0
    this.e = new Exception(EMBED(EMBED const e4c_exception_type_t* END) &RuntimeException; END, "fail", 0, "")
    this.check = false
  end
end
fun deep_nested_rethrow_aux(current : int, a : borrowed Acc) : unit
  try
    if current > 0 then
      deep_nested_rethrow_aux(current-1,a)
    else
      try
        throw RuntimeException("deepest")
      catch e : RuntimeException
        a.e = e
        a.check = true
        e.rethrow()
      end
      ()
    end
  catch e : RuntimeException
    a.res = a.res + 1
    a.check = a.check && a.e == e
    e.rethrow()
    ()
  end
end
fun test_deep_nested_rethrow() : unit
  var a = new Acc()
  try
    deep_nested_rethrow_aux(4,a)
  catch e : RuntimeException
    ()
  end
  println("{} {} {}", a.res, a.check, a.e.message)
end

fun test_gete_right(x : int) : unit
  var either = Right[Exception, int](x)
  try
    println(gete[int](either))
  catch
    println(-x)
  end
end

fun test_gete_left(x : int) : unit
  -- Produce Either with result on right or Exception on left.
  var either =
    try
      throw RuntimeException
      Right[Exception, int](-x)
    catch exc : RuntimeException
      Left[Exception, int](exc)
    end
  try
    gete[int](either)
    println(-x)
  catch
    println(x)
  end
end

active class Main
  def main() : unit
    test_tryCatch(1)
    test_tryCatch_throw(2)
    test_tryCatch_explicit(3)
    test_tryCatch_throw_explicit(4)
    test_throw_without_msg(5)
    test_tryFin(6)
    test_tryCatchFin(7)
    test_tryCatchFin_throw(8)
    test_tryCatchFin_nested(9)
    test_tryCatchFin_nested_throw(10)
    test_tryFin_throw(11)
    test_exception_var(12)
    test_exception_msg()
    test_exception_info()
    test_catch_custom_throw(15)
    test_catch_supertype(16)
    test_multiple_catch_branches()
    test_multiple_catch_branches2()
    test_passive_classes_throw()
    test_rethrow()
    test_gete_right(21)
    test_gete_left(22)
    test_deep_nesting()
    test_nested_rethrow()
    test_deep_nested_rethrow()
  end
end

