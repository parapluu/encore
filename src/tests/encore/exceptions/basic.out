1
2
3
4
5
600
700
800
900
1000
1100
12
informative_msg
type:RuntimeException, message:informative_msg, line:NOT_TESTING, file:NOT_TESTING
15
16
passed_multiple_catch_branches
passed_multiple_catch_branches2
caught LocalException, ReadException, LinearException
test_rethrow: pass
21
22
5050
31 true 4
5 true deepest
